rpc_util.go

```diff
@@ -867,19 +876,44 @@ func recvAndDecompress(p *parser, s recvCompressor, dc Decompressor, maxReceiveM
 }
 
 // Using compressor, decompress d, returning data and size.
-// Optionally, if data will be over maxReceiveMessageSize, just return the size.
-func decompress(compressor encoding.Compressor, d mem.BufferSlice, maxReceiveMessageSize int, pool mem.BufferPool) (mem.BufferSlice, int, error) {
+// If the decompressed data exceeds maxReceiveMessageSize, it returns errMaxMessageSizeExceeded.
+func decompress(compressor encoding.Compressor, d mem.BufferSlice, maxReceiveMessageSize int, pool mem.BufferPool) (mem.BufferSlice, error) {
 	dcReader, err := compressor.Decompress(d.Reader())
 	if err != nil {
-		return nil, 0, err
+		return nil, err
 	}
 
-	out, err := mem.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1), pool)
+	out, err := mem.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)), pool)
 	if err != nil {
 		out.Free()
-		return nil, 0, err
+		return nil, err
+	}
+
+	if doesReceiveMessageOverflow(out.Len(), maxReceiveMessageSize, dcReader) {
+		out.Free()
+		return nil, errMaxMessageSizeExceeded
```

Comment:
Can this just return the `status.Errorf(...)` instead of returning a magic error that gets converted into that?

(We'd want to make sure this method always returns an error that is appropriate to just return directly from `recvAndDecompress`.)